### PR TITLE
drivers: counter: sam: Add qdec as tc special mode

### LINKED ISSUE #71312: ci: samples/drivers/counter/alarm/sample.drivers.counter.alarm build is failing on sam_e70_xplained/same70q21
**To Reproduce**
```
west build -p -b sam_e70_xplained/same70q21 samples/drivers/counter/alarm -T sample.drivers.counter.alarm
```
**Logs and console output**
```
-- west build: making build dir ~/zephyrproject/zephyr/build pristine
-- west build: generating a build system
Loading Zephyr default modules (Zephyr base).
-- Application: ~/zephyrproject/zephyr/samples/drivers/counter/alarm
-- CMake version: 3.22.1
-- Found Python3:~/zephyrproject/.venv/bin/python3 (found suitable version "3.10.12", minimum required is "3.8") found components: Interpreter 
-- Cache files will be written to: ~/.cache/zephyr
-- Zephyr version: 3.6.99 (~/zephyrproject/zephyr)
-- Found west (found suitable version "1.2.0", minimum required is "0.14.0")
-- Board: sam_e70_xplained, qualifiers: same70q21
-- ZEPHYR_TOOLCHAIN_VARIANT not set, trying to locate Zephyr SDK
-- Found host-tools: zephyr 0.16.5 (~/zephyr-sdk-0.16.5)
-- Found toolchain: zephyr 0.16.5 (~/zephyr-sdk-0.16.5)
-- Found Dtc: ~/zephyr-sdk-0.16.5/sysroots/x86_64-pokysdk-linux/usr/bin/dtc (found suitable version "1.6.0", minimum required is "1.4.6") 
-- Found BOARD.dts: ~/zephyrproject/zephyr/boards/atmel/sam/sam_e70_xplained/sam_e70_xplained_same70q21.dts
-- Found devicetree overlay: ~/zephyrproject/zephyr/samples/drivers/counter/alarm/boards/sam_e70_xplained_same70q21.overlay
devicetree error: 'clk' appears in /soc/tc@4000c000 in ~/zephyrproject/zephyr/build/zephyr/zephyr.dts.pre, but is not declared in 'properties:' in ~/zephyrproject/zephyr/dts/bindings/sensor/atmel,sam-tc-qdec.yaml
-- In: ~/zephyrproject/zephyr/build/zephyr, command: ~/zephyrproject/.venv/bin/python3;~/zephyrproject/zephyr/scripts/dts/gen_defines.py;--dts;~/zephyrproject/zephyr/build/zephyr/zephyr.dts.pre;--dtc-flags;'-Wno-unique_unit_address_if_enabled';--bindings-dirs;~/zephyrproject/zephyr/dts/bindings;--header-out;~/zephyrproject/zephyr/build/zephyr/include/generated/devicetree_generated.h.new;--dts-out;~/zephyrproject/zephyr/build/zephyr/zephyr.dts.new;--edt-pickle-out;~/zephyrproject/zephyr/build/zephyr/edt.pickle;--vendor-prefixes;~/zephyrproject/zephyr/dts/bindings/vendor-prefixes.txt
CMake Error at ~/zephyrproject/zephyr/cmake/modules/dts.cmake:295 (message):
  gen_defines.py failed with return code: 1
Call Stack (most recent call first):
  ~/zephyrproject/zephyr/cmake/modules/zephyr_default.cmake:132 (include)
  ~/zephyrproject/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:66 (include)
  ~/zephyrproject/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:92 (include_boilerplate)
  CMakeLists.txt:4 (find_package)

-- Configuring incomplete, errors occurred!
FATAL ERROR: command exited with status 1: /usr/bin/cmake -DWEST_PYTHON=~/zephyrproject/.venv/bin/python3 -B~/zephyrproject/zephyr/build -GNinja -DBOARD=sam_e70_xplained/same70q21 -S~/zephyrproject/zephyr/samples/drivers/counter/alarm
```
**Impact**
blocking CI in https://github.com/zephyrproject-rtos/zephyr/pull/71208

**Environment**
main 731f8d473610f04ec3d0cf53cb10dd373b467f31

### PR DESCRIPTION
**Pull request by nandojve**
The current atmel,sam-tc-qdec sensor implementation shared the timer counter node. This create issues when users wants define both modes. The current proposal changes the qdec dedinition to be a child of tc and refactor all the chain of definitions.

Fixes #71312

### CODE REVIEW

**Review by pdgendt — DISMISSED** [review]

**Comment by nandojve** [comment]
Hi @tswaehn ,

Please, report your tests results here, in the PR. After your tests let me know if it will be necessary add an entry at `samples/sensor/qdec/sample.yaml` to run at CI each time.

**Comment by tswaehn** [comment]
I tested the qdec0 and qdec3 

added to my overlay: 

```
&tc0 {
	status = "okay";

	qdec0:qdec {
		status = "okay";
		label = "QDEC_0";
	};

};

&tc3 {
	status = "okay";

	qdec3: qdec {
		status = "okay";
		label = "QDEC_3";
	};

};
```
and finally used it (as before)

```
devX = device_get_binding("QDEC_0");
devY = device_get_binding("QDEC_3");

sensor_sample_fetch(devX);
struct sensor_value valueX = {0};
sensor_channel_get(devX, SENSOR_CHAN_ROTATION, &valueX);

ret = sensor_sample_fetch(devY);
struct sensor_value valueY = {0};
sensor_channel_get(devY, SENSOR_CHAN_ROTATION, &valueY);
```
looks like it works :)

**Comment by tswaehn** [comment]
this seems to fix board sam_e70_xplained qualifier same70q21

however, there seems to be something missing for the 
board sam_e70_xplained qualifier same70q21b

**Review by pdgendt — DISMISSED** [review]
Could this potentially break OOT usages? Maybe add a migration entry.

**Comment by nandojve** [comment]
Hi @pdgendt ,

> Could this potentially break OOT usages? Maybe add a migration entry.

I don't think OOT is a problem. The driver was kept as it is and the changes are minimal in the DTS.
At release I'll add the entries about the relevant changes.

**Comment by nandojve** [comment]
Hi @tswaehn 

> @nandojve please let me know if I should add it to `same70q21b` as well (or probably you could do?)

Everything is fine now and QDEC is now covered by CI.

**Review by pdgendt — DISMISSED** [review]

File: samples/sensor/qdec/sample.yaml
Comment (MaureenHelm) [inline samples/sensor/qdec/sample.yaml:13]: Why is this build_only?

Suggest renaming the test to `sample.sensor.sam_qdec_sensor` to follow the convention of other qdec tests.

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: samples/sensor/qdec/sample.yaml
Comment (nandojve) [reply samples/sensor/qdec/sample.yaml:13]: Hi @MaureenHelm 

> Why is this build_only?

I think there is no fixture to run this on HW. I can drop if it is not a problem on CI.

> Suggest renaming the test to `sample.sensor.sam_qdec_sensor` to follow the convention of other qdec tests.

I kept the name generic because I assume that any hardware that have this functionality and don't have fixture can be checked at CI using this generic entry. Please, signal to me if I should keep generic or change to your suggestion : )

Comment (nandojve) [reply samples/sensor/qdec/sample.yaml]: I relaxed the list to any that have the alias QDEC0 and ran locally without errors. Let's see the output from CI .
It was necessary keep a special entry for SAM because boards got skipped since they are not `integration_platforms` I presume.

**Review by ubieda — APPROVED** [review]

**Review by pdgendt — APPROVED** [review]

**Comment by aescolar** [comment]
Bulk changing the milestone from everything which did not make it to the v3.7.0 freeze deadline
and which is not labeled as a bug or documentation (or does not appear clearly as only containing
a bug fix or a documentation change).
If you disagree about this change please add the milestone again.

Please note that until rc2 only PRs containing bug fixes, docs and tests can be merged.
After that and until release only bug fixes and docs changes.

Exceptions require approval by the release team and a vote by the TSC.
https://docs.zephyrproject.org/latest/project/release_process.html#stabilization-phase

File: samples/sensor/qdec/sample.yaml
Comment (nandojve) [reply samples/sensor/qdec/sample.yaml]: Hi @MaureenHelm ,

CI is green, I wondering if still anything missing.

**Review by MaureenHelm — APPROVED** [review]

**Comment by nordic-piks** [comment]
@nandojve Changes in sample.yaml introduced problem ie.
it was removed `harness: console` which means that default harness is used, which is ztest. This sample does not produce ztest output.
https://github.com/zephyrproject-rtos/zephyr/pull/73619/files#diff-737a2b2b6eeace96f56ac45f3fdfbdfb9fa049fc3bd89116fd07eb670d31b6c2L7
Needs to be reworked.

**Comment by nandojve** [comment]
Hi @nordic-piks ,

> @nandojve Changes in sample.yaml introduced problem ie. it was removed `harness: console` which means that default harness is used, which is ztest. This sample does not produce ztest output. https://github.com/zephyrproject-rtos/zephyr/pull/73619/files#diff-737a2b2b6eeace96f56ac45f3fdfbdfb9fa049fc3bd89116fd07eb670d31b6c2L7 Needs to be reworked.

I see. Do you know why this not failed on CI then?

**Comment by nordic-piks** [comment]
> Hi @nordic-piks ,
> 
> > @nandojve Changes in sample.yaml introduced problem ie. it was removed `harness: console` which means that default harness is used, which is ztest. This sample does not produce ztest output. https://github.com/zephyrproject-rtos/zephyr/pull/73619/files#diff-737a2b2b6eeace96f56ac45f3fdfbdfb9fa049fc3bd89116fd07eb670d31b6c2L7 Needs to be reworked.
> 
> I see. Do you know why this not failed on CI then?

This will be only visible at HW - when logs from device are parsed.
I think that zephyr CI runs only tests at PC.

**Pull request merged**
